### PR TITLE
Deprecate random.shuffle() and implement random.permutation() for multi-dim inputs

### PR DESCRIPTION
**Pull request by jakevdp**
Addresses #2066. 

Motivation for deprecating `shuffle`:

- the numpy implementation shuffles in-place, which JAX cannot emulate
- numpy's ``permutation`` provides all the functionality of ``shuffle``, but returns a copy rather than shuffling in-place.

Once a few releases have passed, we should make shuffle() raise a NotImplementedError that points to permutation as a viable replacement.

### CODE REVIEW

**Review by froystig — APPROVED** [review]

**Comment by jekbradbury** [comment]
The core team's convention is that we merge our own PRs (after approval). I think you have commit rights now, so feel free to go ahead!

**Comment by jakevdp** [comment]
Thanks!

**Pull request merged**
